qt/widgets/common/src/FunctionModel.cpp

```diff
@@ -44,40 +44,50 @@ void FunctionModel::setFunction(IFunction_sptr fun) {
 }
 
 IFunction_sptr FunctionModel::getFitFunction() const {
-  if (!m_function) {
+  if (!m_function)
     return m_function;
+
+  auto const numberOfFunctions = m_function->nFunctions();
+
+  if (numberOfFunctions > 1) {
+    if (m_currentDomainIndex < numberOfFunctions)
+      return getFitFunctionWithGlobals(m_currentDomainIndex);
+    return getFitFunctionWithGlobals(0);
+
+  } else if (numberOfFunctions == 1) {
+    auto const function = m_function->getFunction(0);
+    auto const composite =
+        std::dynamic_pointer_cast<CompositeFunction>(function);
+
+    if (composite && composite->nFunctions() == 1)
+      return composite->getFunction(0);
+    return function;
   }
-  auto const nf = m_function->nFunctions();
-  if (nf > 1) {
-    auto fun =
-        std::dynamic_pointer_cast<MultiDomainFunction>(m_function->clone());
-
-    auto const singleFun = m_function->getFunction(0);
-    for (auto par = m_globalParameterNames.begin();
-         par != m_globalParameterNames.end();) {
-      if (singleFun->hasParameter(par->toStdString())) {
-        QStringList ties;
-        for (size_t i = 1; i < nf; ++i) {
-          ties << "f" + QString::number(i) + "." + *par;
-        }
-        ties << "f0." + *par;
-        fun->addTies(ties.join("=").toStdString());
-        ++par;
-      } else {
-        par = m_globalParameterNames.erase(par);
-      }
-    }
-    return fun;
-  }
-  if (nf == 1) {
-    auto fun = m_function->getFunction(0);
-    auto compFun = std::dynamic_pointer_cast<CompositeFunction>(fun);
-    if (compFun && compFun->nFunctions() == 1) {
-      return compFun->getFunction(0);
+  return IFunction_sptr();
+}
+
+IFunction_sptr
+FunctionModel::getFitFunctionWithGlobals(std::size_t const &index) const {
+  auto function =
```

Comment:
I realise this was existing code, but I'm wondering why we to make a copy of the function here using ->clone(), whereas if it’s a composite function we just return a pointer to that exact function (and not a copy).

Reply:
I think a copy is made because it adds ties to this function which it doesn't want to be stored in `m_function`. I am not sure why it returns a pointer if its a composite function. It would probably be better if a copy of the composite was returned, I'll make this change 👍